### PR TITLE
update d3.d.ts

### PR DESCRIPTION
**Pull request by ztsai**
update attr signature according to http://bost.ocks.org/mike/nest/ ("You can also access the row index (j) by adding a third argument to the function")

### CODE REVIEW

**Comment by ztsai** [comment]
definition change required in this CR: https://github.com/palantir/plottable/pull/2484

**Comment by vvakame** [comment]
*d3/d3.d.ts*

to authors(@gustavderdrache @borisyankov). could you review this PR?
:+1: or :-1:?

check list

* [ ] pass the Travic-CI test?

**Comment by gustavderdrache** [comment]
Does this apply to the other selection methods? If so, I'd like to see that updated here.

One small point: the `innerIndex` paramter doesn't need to be marked optional. It's not an error for a callback to ignore some of the arguments passed to it.

**Comment by jtlan** [comment]
@gustavderdrache The D3 documentation on this seems pretty slim; [This page](http://bost.ocks.org/mike/nest/) appears to be the only "official" source that mentions the inner index. That said, it seems that at least `style()` and `each()` also accept three arguments (and I would assume the alternate signatures for `attr()` do too).

**Comment by gustavderdrache** [comment]
Yeah... when d3's documentation breaks down in cases like this, I usually go rooting around in the source code. Here's what I found:

Looks like [`each()`](https://github.com/mbostock/d3/blob/master/src/selection/each.js) calls with a node, and two index parameters. And things like [`classed()`](https://github.com/mbostock/d3/blob/master/src/selection/classed.js#L26), [`text()`](https://github.com/mbostock/d3/blob/master/src/selection/text.js#L5), and [`attr()`](https://github.com/mbostock/d3/blob/master/src/selection/attr.js#L19-L23) defer to `each()`.

Transitions are a similar story. For instance, [`attr()`](https://github.com/mbostock/d3/blob/master/src/transition/attr.js#L42) uses [`tween()`](https://github.com/mbostock/d3/blob/master/src/transition/tween.js), and `tween()` ends up calling each, which [passes two index parameters](https://github.com/mbostock/d3/blob/master/src/transition/each.js#L13).

Since every setter ultimately boils down to `.each()`, I vote that they all have support for the third callback argument -- which means both the `foo(name, value)` and `foo({ name: value })` variations. It'll be a lot of work, and it's probably worth adding to the test suite, but the typings will be that much more accurate.

File: d3/d3.d.ts [hunk 1]
```diff
@@ -74,9 +74,9 @@ declare module d3 {
              * Derive an attribute value for each node in the selection based on bound data.
              *
              * @param name The attribute name, optionally prefixed.
-             * @param value The function of the datum (the bound data item) and index (the position in the subgrouping) which computes the attribute value. If the function returns null, the attribute is removed.
+             * @param value The function of the datum (the bound data item), index (the position in the subgrouping), and inner index (overall position in nested selections) which computes the attribute value. If the function returns null, the attribute is removed.
              */
-            attr(name: string, value: (datum: Datum, index: number) => Primitive): Update<Datum>;
+            attr(name: string, value: (datum: Datum, index: number, innerIndex?: number) => Primitive): Update<Datum>;
```
Comment (jtlan) [inline d3/d3.d.ts]: Based on http://bost.ocks.org/mike/nest/, it looks like the last argument should be "`outerIndex`" instead of "`innerIndex`".

**Comment by gustavderdrache** [comment]
:+1: Looks good. Thanks for putting in the extra work on this one.

**Comment by vvakame** [comment]
@gustavderdrache thanks!
@ztsai thanks mate!

**Pull request merged**